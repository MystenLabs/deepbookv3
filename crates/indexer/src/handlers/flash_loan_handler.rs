use crate::handlers::{is_deepbook_tx, struct_tag, try_extract_move_call_package};
use crate::models::deepbook::vault::FlashLoanBorrowed;
use async_trait::async_trait;
use deepbook_schema::models::Flashloan;
use deepbook_schema::schema::flashloans;
use diesel_async::RunQueryDsl;
use move_core_types::account_address::AccountAddress;
use move_core_types::language_storage::StructTag;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::full_checkpoint_content::CheckpointData;
use tracing::debug;

pub struct FlashLoanHandler {
    event_type: StructTag,
}

impl FlashLoanHandler {
    pub fn new(package_id_override: Option<AccountAddress>) -> Self {
        Self {
            event_type: struct_tag::<FlashLoanBorrowed>(package_id_override),
        }
    }
}

impl Processor for FlashLoanHandler {
    const NAME: &'static str = "FlashLoan";
    type Value = Flashloan;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                if !is_deepbook_tx(tx) {
                    return Ok(result);
                }
                let Some(events) = &tx.events else {
                    return Ok(result);
                };

                let package = try_extract_move_call_package(tx).unwrap_or_default();
                let checkpoint_timestamp_ms = checkpoint.checkpoint_summary.timestamp_ms as i64;
                let checkpoint = checkpoint.checkpoint_summary.sequence_number as i64;
                let digest = tx.transaction.digest();

                return events
                    .data
                    .iter()
                    .filter(|ev| ev.type_ == self.event_type)
                    .enumerate()
                    .try_fold(result, |mut result, (index, ev)| {
                        let event: FlashLoanBorrowed = bcs::from_bytes(&ev.contents)?;
                        let data = Flashloan {
                            digest: digest.to_string(),
                            event_digest: format!("{digest}{index}"),
                            sender: tx.transaction.sender_address().to_string(),
                            checkpoint,
                            checkpoint_timestamp_ms,
                            package: package.clone(),
                            pool_id: event.pool_id.to_string(),
                            borrow_quantity: event.borrow_quantity as i64,
                            borrow: true,
                            type_name: event.type_name.to_string(),
                        };
                        debug!("Observed Deepbook Flash Loan Borrowed {:?}", data);
                        result.push(data);
                        Ok(result)
                    });
            })
    }
}

#[async_trait]
impl Handler for FlashLoanHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        Ok(diesel::insert_into(flashloans::table)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?)
    }
}
