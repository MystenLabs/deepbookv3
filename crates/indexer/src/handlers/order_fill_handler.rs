use crate::handlers::{is_deepbook_tx, struct_tag, try_extract_move_call_package};
use crate::models::deepbook::order_info::OrderFilled;
use async_trait::async_trait;
use deepbook_schema::models::OrderFill;
use deepbook_schema::schema::order_fills;
use diesel_async::RunQueryDsl;
use move_core_types::account_address::AccountAddress;
use move_core_types::language_storage::StructTag;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::full_checkpoint_content::CheckpointData;
use tracing::debug;

pub struct OrderFillHandler {
    event_type: StructTag,
}

impl OrderFillHandler {
    pub fn new(package_id_override: Option<AccountAddress>) -> Self {
        Self {
            event_type: struct_tag::<OrderFilled>(package_id_override),
        }
    }
}

impl Processor for OrderFillHandler {
    const NAME: &'static str = "OrderFill";
    type Value = OrderFill;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                if !is_deepbook_tx(tx) {
                    return Ok(result);
                }
                let Some(events) = &tx.events else {
                    return Ok(result);
                };

                let package = try_extract_move_call_package(tx).unwrap_or_default();
                let checkpoint_timestamp_ms = checkpoint.checkpoint_summary.timestamp_ms as i64;
                let checkpoint = checkpoint.checkpoint_summary.sequence_number as i64;
                let digest = tx.transaction.digest();

                return events
                    .data
                    .iter()
                    .filter(|ev| ev.type_ == self.event_type)
                    .enumerate()
                    .try_fold(result, |mut result, (index, ev)| {
                        let event: OrderFilled = bcs::from_bytes(&ev.contents)?;
                        let data = OrderFill {
                            digest: digest.to_string(),
                            event_digest: format!("{digest}{index}"),
                            sender: tx.transaction.sender_address().to_string(),
                            checkpoint,
                            checkpoint_timestamp_ms,
                            package: package.clone(),
                            pool_id: event.pool_id.to_string(),
                            maker_order_id: event.maker_order_id.to_string(),
                            taker_order_id: event.taker_order_id.to_string(),
                            maker_client_order_id: event.maker_client_order_id as i64,
                            taker_client_order_id: event.taker_client_order_id as i64,
                            price: event.price as i64,
                            taker_is_bid: event.taker_is_bid,
                            taker_fee: event.taker_fee as i64,
                            taker_fee_is_deep: event.taker_fee_is_deep,
                            maker_fee: event.maker_fee as i64,
                            maker_fee_is_deep: event.maker_fee_is_deep,
                            base_quantity: event.base_quantity as i64,
                            quote_quantity: event.quote_quantity as i64,
                            maker_balance_manager_id: event.maker_balance_manager_id.to_string(),
                            taker_balance_manager_id: event.taker_balance_manager_id.to_string(),
                            onchain_timestamp: event.timestamp as i64,
                        };
                        debug!("Observed Deepbook Order Filled {:?}", data);
                        result.push(data);
                        Ok(result)
                    });
            })
    }
}

#[async_trait]
impl Handler for OrderFillHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        Ok(diesel::insert_into(order_fills::table)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?)
    }
}
