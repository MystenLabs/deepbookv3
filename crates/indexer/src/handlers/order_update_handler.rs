use crate::handlers::{is_deepbook_tx, struct_tag, try_extract_move_call_package};
use crate::models::deepbook::order::{OrderCanceled, OrderModified};
use crate::models::deepbook::order_info::{OrderExpired, OrderPlaced};
use deepbook_schema::models::{OrderUpdate, OrderUpdateStatus};
use deepbook_schema::schema::order_updates;
use diesel_async::RunQueryDsl;
use move_core_types::account_address::AccountAddress;
use move_core_types::language_storage::StructTag;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::full_checkpoint_content::CheckpointData;
use tracing::debug;

type TransactionMetadata = (String, u64, u64, String, String);

pub struct OrderUpdateHandler {
    order_placed_type: StructTag,
    order_modified_type: StructTag,
    order_canceled_type: StructTag,
    order_expired_type: StructTag,
}

impl OrderUpdateHandler {
    pub fn new(package_id_override: Option<AccountAddress>) -> Self {
        Self {
            order_placed_type: struct_tag::<OrderPlaced>(package_id_override),
            order_modified_type: struct_tag::<OrderModified>(package_id_override),
            order_canceled_type: struct_tag::<OrderCanceled>(package_id_override),
            order_expired_type: struct_tag::<OrderExpired>(package_id_override),
        }
    }
}

impl Processor for OrderUpdateHandler {
    const NAME: &'static str = "OrderUpdate";
    type Value = OrderUpdate;
    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                if !is_deepbook_tx(tx) {
                    return Ok(result);
                }
                let Some(events) = &tx.events else {
                    return Ok(result);
                };

                let package = try_extract_move_call_package(tx).unwrap_or_default();
                let metadata = (
                    tx.transaction.sender_address().to_string(),
                    checkpoint.checkpoint_summary.sequence_number,
                    checkpoint.checkpoint_summary.timestamp_ms,
                    tx.transaction.digest().to_string(),
                    package.clone(),
                );

                return events.data.iter().enumerate().try_fold(
                    result,
                    |mut result, (index, ev)| {
                        if ev.type_ == self.order_placed_type {
                            let event = bcs::from_bytes(&ev.contents)?;
                            result.push(process_order_placed(event, metadata.clone(), index));
                            debug!("Observed Deepbook Order Placed {:?}", tx);
                        } else if ev.type_ == self.order_modified_type {
                            let event = bcs::from_bytes(&ev.contents)?;
                            result.push(process_order_modified(event, metadata.clone(), index));
                            debug!("Observed Deepbook Order Modified {:?}", tx);
                        } else if ev.type_ == self.order_canceled_type {
                            let event = bcs::from_bytes(&ev.contents)?;
                            result.push(process_order_canceled(event, metadata.clone(), index));
                            debug!("Observed Deepbook Order Canceled {:?}", tx);
                        } else if ev.type_ == self.order_expired_type {
                            let event = bcs::from_bytes(&ev.contents)?;
                            result.push(process_order_expired(event, metadata.clone(), index));
                            debug!("Observed Deepbook Order Expired {:?}", tx);
                        }
                        Ok(result)
                    },
                );
            })
    }
}

#[async_trait::async_trait]
impl Handler for OrderUpdateHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        Ok(diesel::insert_into(order_updates::table)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?)
    }
}

fn process_order_placed(
    order_placed: OrderPlaced,
    (sender, checkpoint, checkpoint_timestamp_ms, digest, package): TransactionMetadata,
    event_index: usize,
) -> OrderUpdate {
    let event_digest = format!("{digest}{event_index}");
    OrderUpdate {
        event_digest,
        digest,
        sender,
        checkpoint: checkpoint as i64,
        checkpoint_timestamp_ms: checkpoint_timestamp_ms as i64,
        package,
        status: OrderUpdateStatus::Placed,
        pool_id: order_placed.pool_id.to_string(),
        order_id: order_placed.order_id.to_string(),
        client_order_id: order_placed.client_order_id as i64,
        price: order_placed.price as i64,
        is_bid: order_placed.is_bid,
        onchain_timestamp: order_placed.timestamp as i64,
        original_quantity: order_placed.placed_quantity as i64,
        quantity: order_placed.placed_quantity as i64,
        filled_quantity: 0,
        trader: order_placed.trader.to_string(),
        balance_manager_id: order_placed.balance_manager_id.to_string(),
    }
}

fn process_order_modified(
    order_modified: OrderModified,
    (sender, checkpoint, checkpoint_timestamp_ms, digest, package): TransactionMetadata,
    event_index: usize,
) -> OrderUpdate {
    let event_digest = format!("{digest}{event_index}");
    OrderUpdate {
        digest,
        event_digest,
        sender,
        checkpoint: checkpoint as i64,
        checkpoint_timestamp_ms: checkpoint_timestamp_ms as i64,
        package,
        status: OrderUpdateStatus::Modified,
        pool_id: order_modified.pool_id.to_string(),
        order_id: order_modified.order_id.to_string(),
        client_order_id: order_modified.client_order_id as i64,
        price: order_modified.price as i64,
        is_bid: order_modified.is_bid,
        onchain_timestamp: order_modified.timestamp as i64,
        original_quantity: order_modified.previous_quantity as i64,
        quantity: order_modified.new_quantity as i64,
        filled_quantity: order_modified.filled_quantity as i64,
        trader: order_modified.trader.to_string(),
        balance_manager_id: order_modified.balance_manager_id.to_string(),
    }
}

fn process_order_canceled(
    order_canceled: OrderCanceled,
    (sender, checkpoint, checkpoint_timestamp_ms, digest, package): TransactionMetadata,
    event_index: usize,
) -> OrderUpdate {
    let event_digest = format!("{digest}{event_index}");
    OrderUpdate {
        digest,
        event_digest,
        sender,
        checkpoint: checkpoint as i64,
        checkpoint_timestamp_ms: checkpoint_timestamp_ms as i64,
        package,
        status: OrderUpdateStatus::Canceled,
        pool_id: order_canceled.pool_id.to_string(),
        order_id: order_canceled.order_id.to_string(),
        client_order_id: order_canceled.client_order_id as i64,
        price: order_canceled.price as i64,
        is_bid: order_canceled.is_bid,
        onchain_timestamp: order_canceled.timestamp as i64,
        original_quantity: order_canceled.original_quantity as i64,
        quantity: order_canceled.base_asset_quantity_canceled as i64,
        filled_quantity: (order_canceled.original_quantity
            - order_canceled.base_asset_quantity_canceled) as i64,
        trader: order_canceled.trader.to_string(),
        balance_manager_id: order_canceled.balance_manager_id.to_string(),
    }
}

fn process_order_expired(
    order_expired: OrderExpired,
    (sender, checkpoint, checkpoint_timestamp_ms, digest, package): TransactionMetadata,
    event_index: usize,
) -> OrderUpdate {
    let event_digest = format!("{digest}{event_index}");
    OrderUpdate {
        digest,
        event_digest,
        sender,
        checkpoint: checkpoint as i64,
        checkpoint_timestamp_ms: checkpoint_timestamp_ms as i64,
        package,
        status: OrderUpdateStatus::Expired,
        pool_id: order_expired.pool_id.to_string(),
        order_id: order_expired.order_id.to_string(),
        client_order_id: order_expired.client_order_id as i64,
        price: order_expired.price as i64,
        is_bid: order_expired.is_bid,
        onchain_timestamp: order_expired.timestamp as i64,
        original_quantity: order_expired.original_quantity as i64,
        quantity: order_expired.base_asset_quantity_canceled as i64,
        filled_quantity: (order_expired.original_quantity
            - order_expired.base_asset_quantity_canceled) as i64,
        trader: order_expired.trader.to_string(),
        balance_manager_id: order_expired.balance_manager_id.to_string(),
    }
}
