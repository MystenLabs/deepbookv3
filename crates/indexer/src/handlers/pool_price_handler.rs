use crate::handlers::{is_deepbook_tx, struct_tag, try_extract_move_call_package};
use crate::models::deepbook::deep_price::PriceAdded;
use async_trait::async_trait;
use deepbook_schema::models::PoolPrice;
use deepbook_schema::schema::pool_prices;
use diesel_async::RunQueryDsl;
use move_core_types::account_address::AccountAddress;
use move_core_types::language_storage::StructTag;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::full_checkpoint_content::CheckpointData;
use tracing::debug;

pub struct PoolPriceHandler {
    event_type: StructTag,
}

impl PoolPriceHandler {
    pub fn new(package_id_override: Option<AccountAddress>) -> Self {
        Self {
            event_type: struct_tag::<PriceAdded>(package_id_override),
        }
    }
}

impl Processor for PoolPriceHandler {
    const NAME: &'static str = "PoolPrice";
    type Value = PoolPrice;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                if !is_deepbook_tx(tx) {
                    return Ok(result);
                }
                let Some(events) = &tx.events else {
                    return Ok(result);
                };

                let package = try_extract_move_call_package(tx).unwrap_or_default();
                let checkpoint_timestamp_ms = checkpoint.checkpoint_summary.timestamp_ms as i64;
                let checkpoint = checkpoint.checkpoint_summary.sequence_number as i64;
                let digest = tx.transaction.digest();

                return events
                    .data
                    .iter()
                    .filter(|ev| ev.type_ == self.event_type)
                    .enumerate()
                    .try_fold(result, |mut result, (index, ev)| {
                        let event: PriceAdded = bcs::from_bytes(&ev.contents)?;
                        let data = PoolPrice {
                            digest: digest.to_string(),
                            event_digest: format!("{digest}{index}"),
                            sender: tx.transaction.sender_address().to_string(),
                            checkpoint,
                            checkpoint_timestamp_ms,
                            package: package.clone(),
                            target_pool: event.target_pool.to_string(),
                            conversion_rate: event.conversion_rate as i64,
                            reference_pool: event.reference_pool.to_string(),
                        };
                        debug!("Observed Deepbook Price Addition {:?}", data);
                        result.push(data);
                        Ok(result)
                    });
            })
    }
}

#[async_trait]
impl Handler for PoolPriceHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        Ok(diesel::insert_into(pool_prices::table)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?)
    }
}
