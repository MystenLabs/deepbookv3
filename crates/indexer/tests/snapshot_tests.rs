use chrono::NaiveDateTime;
use deepbook_indexer::handlers::balances_handler::BalancesHandler;
use deepbook_indexer::handlers::flash_loan_handler::FlashLoanHandler;
use deepbook_indexer::handlers::order_fill_handler::OrderFillHandler;
use deepbook_indexer::handlers::order_update_handler::OrderUpdateHandler;
use deepbook_indexer::DeepbookEnv;
use deepbook_schema::MIGRATIONS;
use fastcrypto::hash::{HashFunction, Sha256};
use insta::assert_json_snapshot;
use serde_json::Value;
use sqlx::{Column, PgPool, Row, ValueRef};
use std::fs;
use std::path::Path;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_indexer_alt_framework::store::Store;
use sui_pg_db::temp::TempDb;
use sui_pg_db::Connection;
use sui_pg_db::Db;
use sui_pg_db::DbArgs;
use sui_storage::blob::Blob;
use sui_types::full_checkpoint_content::CheckpointData;
use deepbook_indexer::handlers::pool_price_handler::PoolPriceHandler;

#[tokio::test]
async fn balances_test() -> Result<(), anyhow::Error> {
    let handler = BalancesHandler::new(DeepbookEnv::Mainnet);
    data_test("balances", handler, ["balances"]).await?;
    Ok(())
}

#[tokio::test]
async fn flash_loan_test() -> Result<(), anyhow::Error> {
    let handler = FlashLoanHandler::new(DeepbookEnv::Mainnet);
    data_test("flash_loans", handler, ["flashloans"]).await?;
    Ok(())
}

#[tokio::test]
async fn order_fill_test() -> Result<(), anyhow::Error> {
    let handler = OrderFillHandler::new(DeepbookEnv::Mainnet);
    data_test("order_fill", handler, ["order_fills"]).await?;
    Ok(())
}
#[tokio::test]
async fn order_update_test() -> Result<(), anyhow::Error> {
    let handler = OrderUpdateHandler::new(DeepbookEnv::Mainnet);
    data_test("order_update", handler, ["order_updates"]).await?;
    Ok(())
}

#[tokio::test]
async fn pool_price_test() -> Result<(), anyhow::Error> {
    let handler = PoolPriceHandler::new(DeepbookEnv::Mainnet);
    data_test("pool_price", handler, ["pool_prices"]).await?;
    Ok(())
}

async fn data_test<H, I>(
    test_name: &str,
    handler: H,
    tables_to_check: I,
) -> Result<(), anyhow::Error>
where
    I: IntoIterator<Item=&'static str>,
    H: Handler + Processor,
    for<'a> H::Store: Store<Connection<'a>=Connection<'a>>,
{
    // Set up the temporary database
    let temp_db = TempDb::new()?;
    let url = temp_db.database().url();
    let db = Arc::new(Db::for_write(url.clone(), DbArgs::default()).await?);
    db.run_migrations(MIGRATIONS).await?;
    let mut conn = db.connect().await?;

    // Test setup based on provided test_name
    let test_path = Path::new("tests/checkpoints").join(test_name);
    let checkpoints = get_checkpoints_in_folder(&test_path)?;

    // Run pipeline for each checkpoint
    for checkpoint in checkpoints {
        run_pipeline(&handler, &checkpoint, &mut conn).await?;
    }

    // Check results by comparing database tables with snapshots
    for table in tables_to_check {
        let rows = read_table(&table, &url.to_string()).await?;
        assert_json_snapshot!(format!("{test_name}__{table}"), rows);
    }
    Ok(())
}

async fn run_pipeline<'c, T: Handler + Processor, P: AsRef<Path>>(
    handler: &T,
    path: P,
    conn: &mut Connection<'c>,
) -> Result<(), anyhow::Error>
where
    T::Store: Store<Connection<'c>=Connection<'c>>,
{
    let bytes = fs::read(path)?;
    let cp = Blob::from_bytes::<CheckpointData>(&bytes)?;
    let result = handler.process(&Arc::new(cp))?;
    T::commit(&result, conn).await?;
    Ok(())
}

/// Read the entire table from database as json value.
/// note: bytea values will be hashed to reduce output size.
async fn read_table(table_name: &str, db_url: &str) -> Result<Vec<Value>, anyhow::Error> {
    let pool = PgPool::connect(db_url).await?;
    let rows = sqlx::query(&format!("SELECT * FROM {table_name}"))
        .fetch_all(&pool)
        .await?;

    // To json
    Ok(rows
        .iter()
        .map(|row| {
            let mut obj = serde_json::Map::new();

            for column in row.columns() {
                let column_name = column.name();

                // timestamp is the insert time in deepbook DB, hardcoding it to a fix value.
                if column_name == "timestamp" {
                    obj.insert(column_name.to_string(), Value::String("1970-01-01 00:00:00.000000".to_string()));
                    continue;
                }

                let value = if let Ok(v) = row.try_get::<String, _>(column_name) {
                    Value::String(v)
                } else if let Ok(v) = row.try_get::<i32, _>(column_name) {
                    Value::Number(v.into())
                } else if let Ok(v) = row.try_get::<i64, _>(column_name) {
                    Value::Number(v.into())
                } else if let Ok(v) = row.try_get::<bool, _>(column_name) {
                    Value::Bool(v)
                } else if let Ok(v) = row.try_get::<Value, _>(column_name) {
                    v
                } else if let Ok(v) = row.try_get::<Vec<u8>, _>(column_name) {
                    // hash bytea contents
                    let mut hash_function = Sha256::default();
                    hash_function.update(v);
                    let digest2 = hash_function.finalize();
                    Value::String(digest2.to_string())
                } else if let Ok(v) = row.try_get::<NaiveDateTime, _>(column_name) {
                    Value::String(v.to_string())
                } else if let Ok(true) = row.try_get_raw(column_name).map(|v| v.is_null()) {
                    Value::Null
                } else {
                    panic!(
                        "Cannot parse DB value to json, type: {:?}, column: {column_name}",
                        row.try_get_raw(column_name)
                            .map(|v| v.type_info().to_string())
                    )
                };
                obj.insert(column_name.to_string(), value);
            }

            Value::Object(obj)
        })
        .collect())
}

fn get_checkpoints_in_folder(folder: &Path) -> Result<Vec<String>, anyhow::Error> {
    let mut files = Vec::new();

    // Read the directory
    for entry in fs::read_dir(folder)? {
        let entry = entry?;
        let path = entry.path();

        // Check if it's a file and ends with ".chk"
        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("chk") {
            files.push(path.display().to_string());
        }
    }

    Ok(files)
}
