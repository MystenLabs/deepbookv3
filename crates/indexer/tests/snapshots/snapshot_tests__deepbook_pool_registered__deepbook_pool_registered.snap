---
source: crates/indexer/tests/snapshot_tests.rs
expression: rows
---
[
  {
    "event_digest": "FZE9pv8CnWakwaWC1SoYiLUPPPidWXsy3aabScVoUvRz0",
    "digest": "FZE9pv8CnWakwaWC1SoYiLUPPPidWXsy3aabScVoUvRz",
    "sender": "0xb3d277c50f7b846a5f609a8d13428ae482b5826bb98437997373f3a0d60d280e",
    "checkpoint": "248053954",
    "timestamp": "1970-01-01 00:00:00.000000",
    "checkpoint_timestamp_ms": "1759426717281",
    "package": "0x442d21fd044b90274934614c3c41416c83582f42eaa8feb4fecea301aa6bdd54",
    "pool_id": "0x1c19362ca52b8ffd7a33cee805a67d40f31e6ba303753fd3a4cfdfacea7163a5",
    "onchain_timestamp": "1759426717142"
  }
]
